kythe/docs/schema/indexing-protobuf.txt

```diff
@@ -31,12 +31,235 @@ implementation, as this is often difficult (or uninteresting) to read. This
 The current sample web UI does not interpret the parts of the schema we will
 use; this is a work in progress.
 
-== Protocol Buffers
+=== Java To JavaScript with *imputes* edges
+
+This approach is generic and works for any combination of AliceLang and
+BobLang. In this example we generate JavaScript files from Java file so
+AliceLang is Java and BobLang is JavaScript. Given `Color.java`:
+
+[source,java]
+-------------------------------------------------------------------------------
+public enum Color {
+  RED;
+}
+-------------------------------------------------------------------------------
+
+Generator produces `color.js`:
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+-------------------------------------------------------------------------------
+
+==== Changes to Generator
+
+To support cross-references betwen `color.js` and `Color.java` we need to update
+Generator to output the following mapping data for `Color`, `RED` and `GREEN`
+elements.
+
+[source,json]
+-------------------------------------------------------------------------------
+{
+    "type": "kythe0",
+    "meta": [{
+        "type": "anchor_anchor",
+        "source_begin": 13,
+        "source_end": 18,
+        "target_begin": 6,
+        "target_end": 11,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }, {
+        "type": "anchor_anchor",
+        "source_begin": 22,
+        "source_end": 25,
+        "target_begin": 18,
+        "target_end": 21,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }]
+}
+-------------------------------------------------------------------------------
+
+This mapping has 3 `meta` entries. The first entry for `Color`, the second for
+`RED`. Note:
+
+* Each entry doesn't contain names of elements. Each entry contains only
+  position of elements in the source (`Color.java`) and target (`color.js`)
+  files.
+* Each position is defined as byte offset inside file and not as line/column.
+  This is required because in Kythe anchors defined using byte offsets and
+  not line/column. In this example JavaScript indexer will process this
+  mapping and will need to output *anchor* for `Color.java` and indexer
+  doesn't have access to the `Color.java` file (it has access only to JS
+  files). Because of that JS indexer can't translate line/column to byte
+  offset.
+* Entry doesn't contain vnames of elements in `Color.java` or `color.js` and
+  instead contains positions. Vnames of nodes are internal details of each
+  indexer and subject to change. Generator usually a standalone tool that
+  doesn't know rules for producing vnames for specific language so it's
+  impossible for Generator to output vnames of nodes.
```

Comment:
-ish

For things like protobuf (and mojom?) the VName generation rules are specified to make it possible to create generates edges, rather than relying on imputes.  It might be worth expanding on that a bit? (or not, that might just make it confusing).

Reply:
How would you suggest to expand? There is a whole section protobuf below. I can add something like "for another approach based on VNames see Protcol Buffers section below"?

Reply:
I think a parenthetical to that effect would be useful, yeah.

Reply:
Added a sentence.